libsignal-protocol/src/keys/public.rs

```diff
@@ -62,6 +63,32 @@ impl PublicKey {
             }
         }
     }
+
+    /// Uses this public key to check the ECDH agreement with a private key
+    pub fn calculate_agreement(
+        &self,
+        private_key: PrivateKey,
+    ) -> Result<Vec<u8>, Error> {
+        unsafe {
+            let mut shared_data = std::ptr::null_mut();
+            let length = sys::curve_calculate_agreement(
+                &mut shared_data,
+                self.raw.as_const_ptr(),
+                private_key.raw.as_const_ptr(),
+            ) as usize;
+            if length > 0 {
+                // FIXME: this only works because by default on Linux, Rust uses the
```

Comment:
Instead of relying on a platform-specific quirk, why not just copy the `secret` to a `Vec` and call `libc::free()` to delete the original data?

Maybe something like this:

```rust
unsafe {
  let mut shared_data = std::ptr::null_mut();
  let length = sys::curve_calculate_agreement(
                &mut shared_data,
                self.raw.as_const_ptr(),
                private_key.raw.as_const_ptr(),
            ) as usize;

  let secret = Vec::with_capacity(length); 
  std::ptr::copy_nonoverlapping(c_buffer, secret.as_mut_ptr(), length); 
  secret.set_len(length);
  libc::free(c_buffer);
}
```

Reply:
Another way would be to introduce a struct that uses the data from the pointer directly, and implement `Drop` to call `libc::free`. This would avoid the extra allocation and copying, but I don't think we'll ever face performance issues because of this?